subiquity/controllers/filesystem.py

```diff
@@ -478,21 +510,48 @@ def volgroup_handler(self, existing, spec):
 
     def make_boot_disk(self, new_boot_disk):
         boot_partition = None
-        for disk in self.model.all_disks():
-            for part in disk.partitions():
-                if part.flag in ("bios_grub", "boot", "prep"):
-                    boot_partition = part
+        if self.model.bootloader == Bootloader.BIOS:
+            install_dev = self.model.grub_install_device
+            if install_dev:
+                boot_partition = install_dev._potential_boot_partition()
+        elif self.model.bootloader == Bootloader.UEFI:
+            mount = self.model._mount_for_path("/boot/efi")
+            if mount is not None:
+                boot_partition = mount.device.volume
+        elif self.model.bootloader == Bootloader.PREP:
+            boot_partition = self.model.grub_install_device
         if boot_partition is not None:
-            boot_disk = boot_partition.device
-            full = boot_disk.free_for_partitions == 0
-            self.delete_partition(boot_partition)
-            if full:
-                largest_part = max(
-                    boot_disk.partitions(), key=lambda p: p.size)
-                largest_part.size += boot_partition.size
-            if new_boot_disk.free_for_partitions < boot_partition.size:
-                largest_part = max(
-                    new_boot_disk.partitions(), key=lambda p: p.size)
-                largest_part.size -= (
-                    boot_partition.size - new_boot_disk.free_for_partitions)
-        self._create_boot_partition(new_boot_disk)
+            if boot_partition.preserve:
+                if self.model.bootloader == Bootloader.PREP:
+                    boot_partition.wipe = None
+                elif self.model.bootloader == Bootloader.UEFI:
+                    self.delete_mount(boot_partition.fs().mount())
```

Comment:
Why does this need removing?  Do you create a new mount entry?

Reply:
Yes, the scenario here is: there is a disk with an existing ESP. It has been selected to be active, but now the user is making some other ESP the active one. It needs to be unmounted before we create the mount for the new one.

Reply:
we used to ignore user choices like that, and force reuse ESP. or at least that was my experience in like ubiquity.

Reply:
Well OK, are you saying we should change here? I think the two cases here are: installing to a partition of a disk which already has an ESP, in which case the default is to reuse it, or installing to a separate disk (e.g. a system with a "windows disk" and a "linux disk") in which case creating a new ESP seems a better default to me. The user can then change this (that's what this code is about) but presumably approximately 0% of users will do that.